[#5247] Pro Metrics weekly mailer

lizconlan opened the pull request:
## Relevant issue(s)

Closes #5247 

## What does this do?

* Brings the pro metrics script we've been running manually into the codebase
* Addresses a possible omission from the `:pro_user` factory where it didn't create a `pro_account` for the created user
* Creates an email with the report results
* Adds a cron task to run and email the report [every Monday morning](https://crontab.guru/#37_8_*_*_1)

## Why was this needed?

* We'd like to be able to automate the process of running the pro metrics script and have it send an email to the relevant Pro Team
* We'd like to be able to offer this as a metrics report for reusers who want to use the Pro features (without them needing to run a command line script!)

## Implementation notes

* The specs are not ideal as there are a couple of missing features in StripeMock:
  - it doesn't model the [lifecycle for subscriptions well](https://github.com/rebelidealist/stripe-ruby-mock/issues/286) so I've added a hacky monkeypatch for that for now (I looked to see if it was possible to quickly implement the stale PR against their codebase that attempted to resolve in a more Stripe-like way but it wasn't worth holding up feature development for)
  - it only has a basic implementation of `Stripe::Subscription.list` which does not filter on `plan` or `created[gte]`/`created[lte]` so the specs aren't as comprehensive as I'd like (and we might not need the code that checks the `plan.id` before appending it!)
* There might be an unintended bug - ported from the original - where we only report the number of **embargoed** batches, have left that as-is for reporting consistency
* Returns the same data on the live server as running the original script (some of the counts may differ slightly as the report now runs for a set reporting period based on start/end of day rather than using the script run timings)

### Sample script output

From the new script:
```rb
{:new_pro_requests=>104,
 :total_new_requests=>37535,
 :new_batches=>3,
 :new_signups=>3,
 :total_accounts=>284,
 :active_accounts=>42,
 :paying_users=>74,
 :discounted_users=>17,
 :trialing_users=>81,
 :past_due_users=>{:count=>0, :subs=>0},
 :pending_cancellations=>
  {:count=>2, :subs=>["sub_redacted", "sub_redacted"]},
 :unpaid_users=>{:count=>0, :subs=>0},
 :unknown_users=>0,
 :new_and_returning_users=>
  {:count=>4,
   :subs=>
    ["sub_redacted",
     "sub_redacted",
     "sub_redacted",
     "sub_redacted"]},
 :canceled_users=>{:count=>0, :subs=>[]}}
```

From the old script:
```rb
{"Number of requests made this week"=>104,
 "Estimated number of \"Pro\" requests"=>37535,
 "Number of batch requests made this week"=>3,
 "Number of pro signups this week"=>3,
 "Total number of pro accounts"=>284,
 "Number of pro accounts active this week"=>42,
 "Paying Users"=>74,
 "Discounted Users"=>17,
 "Trialing Users"=>81,
 "Past Due Users"=>{:count=>0, :subs=>0},
 "Canceled Users"=>
  {:count=>2, :subs=>["sub_redacted", "sub_redacted"]},
 "Unpaid Users"=>{:count=>0, :subs=>0},
 "Unknown Users"=>0}
```

## Email output

Preview available at: http://10.10.10.30:3000/rails/mailers/alaveteli_pro/metrics_mailer/weekly_report
(looks different with and without `:pro_pricing` enabled)

### Screenshots
 
#### Without pricing
<img width="526" alt="Screen Shot 2019-07-09 at 18 47 09" src="https://user-images.githubusercontent.com/27760/60952655-8e51f800-a2f3-11e9-8870-516129c67f84.png">

#### With pricing
<img width="633" alt="Screen Shot 2019-07-09 at 18 45 54" src="https://user-images.githubusercontent.com/27760/60952444-34e9c900-a2f3-11e9-86be-54a642f449d0.png">

### Missing specs

~The pending specs are currently the ones around Stripe's billing lifecycle for Subscriptions: https://stripe.com/docs/billing/lifecycle#inactive~

~* Number of past due users (pre-existing)
  [`past_due`](https://stripe.com/docs/billing/lifecycle#automatic-invoice-lifecycle) - a payment fails, the subscription is flagged as `past_due` and is waiting for the next try; status may clear when a subsequent attempt succeeds or go on to become `unpaid`~
~* Number of unpaid users (pre-existing, should not be possible in live data with our current Stripe config)
  `unpaid` - not actually a reachable state for our recommended setup as we automatically cancel the subscription once the maximum number of tries have been exceeded so it should be deleted rather than moving to `unpaid`; we may not need to include this one in reporting at all~

~This [SO thread](https://stackoverflow.com/questions/24815799/how-can-i-test-unpaid-subscriptions-in-stripe-with-ruby-on-rails) - despite being slightly out of date - feels like it should help but hasn't so far.~

There is no spec for unknown users as I think it's a case of "Stripe returned data it has no description for" so there's no real way to model it unless/until it happens; probably a catchall. I have not included it in the emailed report as it's pretty meaningless so we could probably drop this from the reporting data as well (unless it's a useful cross check against something going wrong and we use it as a flag to make some kind of error warning / contact tech support message appear?).

lizconlan commented on lib/alaveteli_pro/metrics_report.rb:
```diff
@@ -0,0 +1,190 @@
+        Stripe::Subscription.retrieve(subscription_id).customer
+      if account = ProAccount.find_by(stripe_customer_id: customer_id)
+        account.user
+      else
```
fixed by using `.try(:user)` instead

lizconlan commented:
Things to document:
- How to set up the Stripe webhooks
  - the endpoint
  - which webhooks to subscribe to
- The 'Automatic collection' payment settings (under Settings > Stripe apps > Billing > 'Automatic collection'), includes the custom behaviours set in 'Manage failed payments' such as 'if all retries fail - 'cancel subscription'/'mark as unpaid'/'leave as-is')
- Question: Do we have to set the API version or do we set that in the code?

lizconlan commented:
> intermediate commits that were a waste of time reviewing

An attempt at showing refactoring of working code into something we'd like to use in the app without a potentially baffling "Changed all the things" commit

garethrees commented:
Yeah, it was just hard to know whether to comment on something that looked like a problem, because it wasn't obvious whether it was going to get fixed in a later commit.

Definitely appreciate the effort in getting it cleaned up, but I think refactoring is clearer starting from existing code. In this scenario we were adding completely new (as in, new to the app, not new to us) code to the app.

Rather than paste it all, it would have been clearer to port it over bit by bit (and then clean up along the way). That way each diff would have been smaller (as per your very reasonable point of not having a giant "everything changed" commit!), but also each would be functioning correctly, rather than the situation here where we've got a mix of working and "to be fixed later" code.

garethrees commented:
p.s. don't forget changelog entry

lizconlan commented:
> p.s. don't forget changelog entry

Added a `Co-authored by:` to f2530b0 to try to reflect the dual authorship of the original code a bit better and added "adapted from a locally run metrics script written by Gareth" to the changelog entry before taking all the credit (blame?)

Not going to get merged until tomorrow as it's still waiting for CI to run so time to change it if that's not quite right

garethrees commented:
Just keep it simple:

```
Send weekly metrics email to the Pro Admin team (Liz Conlan, Gareth Rees)
```
